doc/dev/error-categorization.md

```diff
@@ -0,0 +1,111 @@
+# Error Categorization
+
+In this document we start to explain how to categorize errors.
+Many errors are unique and can be difficult to categorize so this
+document will act as a guideline.
+
+## Categorization Mindset
+
+Errors along with their unique code only exist because they can be **reacted
+differently on programmatically**. Imagine a method call which returns a `Timeout` error. Naturally
+a senseful reaction would be to retry at a later point in time. So
+the reaction here would be to retry with a time based approach. On the other hand
+it does not make sense to differentiate between `No Write Permission` and `No Read Permission` as
+the application just knows that it simply cannot access the desired resource and tells
+the user to grant it. One could argue that an application could always react differently
+based on the error (eg. give yourself write or read permission) but such detailed granularity
+is rarely needed and would come with a lot of maintenance effort. In the previous versions of Elektra
+we were facing many duplicated errors because too many existed and developers simply cannot know each
+and every code.
+
+So if you feel the need for a new category you should first ask yourself the following question:
+> Can I react differently if I would know every error X in category Y?
+
+If not it will most likely fit into an existing category. Categories are leaf based, so you cannot
+put an error into a node such as `Permanent errors` (see below).
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+- Permanent errors ("01000")
+  - Resource ("01100")
+  - Parsing ("01200")
+  - Installation ("01300")
+  - Logical ("01400")
+- Conflict ("02000")
+- Timeout ("03000")
+- Validation ("04000")
+  - Syntactic ("04100")
+  - Semantic ("04200")
+### Permanent errors ("01000")
+
+The main category `Permanent Errors` refer to such errors which cannot be fixed by retry 
```

Comment:
Timeouts are the only ones that might be fixed by retrying without any changes.

Resource errors can definitely be fixed by retrying (after changing the permissions).

Reply:
Should I write "cannot be fixed by a simple retry"? Or what is your suggested solution here?

Reply:
That still doesn't address the fact that "conflict" and "validation" errors also cannot be fixed by a "simple retry".